<launch>
	<!-- vim: set ft=xml noet : -->
	<!-- launch script for PX4 based FCU's -->

    <!-- mavros parameters -->
	<arg name="fcu_conn" default="usb" />
	<arg name="gcs_bridge" default="" />
    <arg name="fcu_ip" default="127.0.0.1"/>
	<arg name="tgt_system" default="1" />
	<arg name="tgt_component" default="1" />
	<arg name="log_output" default="screen" />
	<arg name="fcu_protocol" default="v2.0" />
	<arg name="respawn_mavros" default="false" />
    <arg name="pluginlists_yaml" value="$(find px4_air)/launch/px4_pluginlists.yaml" />
	<arg name="config_yaml" value="$(find px4_air)/launch/px4_config.yaml" />
    <!-- mavros extras parameters -->
    <arg name="viz" default="true"/>

	<node pkg="mavros" type="mavros_node" name="mavros" required="$(eval not respawn_mavros)" clear_params="true" output="$(arg log_output)" respawn="$(arg respawn_mavros)">
        <!-- UART connection -->
        <param name="fcu_url" value="/dev/ttyAMA0:921600" if="$(eval fcu_conn is None or fcu_conn == 'uart')"/>

        <!-- USB connection -->
        <param name="fcu_url" value="/dev/ttyACM0:57600" if="$(eval fcu_conn == 'usb')"/>

        <!-- sitl before PX4 1.9.0 -->
        <param name="fcu_url" value="udp://@$(arg fcu_ip):14557" if="$(eval fcu_conn == 'udp')"/>

        <!-- sitl since PX4 1.9.0 -->
        <param name="fcu_url" value="udp://@$(arg fcu_ip):14580" if="$(eval fcu_conn == 'sitl')"/>

        <!-- hitl connection (to gazebo_mavlink_interface plugin) -->
        <param name="fcu_url" value="udp://$(arg fcu_ip):14540@" if="$(eval fcu_conn == 'hitl')"/>


        <!-- gcs bridge -->
        <param name="gcs_url" value="tcp-l://0.0.0.0:5760" if="$(eval gcs_bridge == 'tcp')"/>
        <param name="gcs_url" value="udp://0.0.0.0:14550@14550" if="$(eval gcs_bridge == 'udp')"/>
        <param name="gcs_url" value="udp-b://$(env ROS_HOSTNAME):14550@14550" if="$(eval gcs_bridge == 'udp-b')"/>
        <param name="gcs_url" value="udp-pb://$(env ROS_HOSTNAME):14550@14550" if="$(eval gcs_bridge == 'udp-pb')"/>
        <param name="gcs_url" value="" if="$(eval not gcs_bridge)"/>
        
		<param name="target_system_id" value="$(arg tgt_system)" />
		<param name="target_component_id" value="$(arg tgt_component)" />
		<param name="fcu_protocol" value="$(arg fcu_protocol)" />

		<!-- load blacklist, config -->
		<rosparam command="load" file="$(arg pluginlists_yaml)" />
		<rosparam command="load" file="$(arg config_yaml)" />
	</node>

    <!-- remapped distance_sensor config -->
    <!-- <rosparam param="$(arg distance_sensor_remap)" if="$(eval bool(distance_sensor_remap))">
        subscriber: true
        id: 1
        orientation: PITCH_270
        covariance: 1  # cm
    </rosparam> -->

    <!-- Rangefinders frame -->
    <node pkg="tf2_ros" type="static_transform_publisher" name="rangefinder_frame" args="0 0 -0.05 0 1.5707963268 0 base_link rangefinder"/>

    <!-- Copter visualization -->
    <node name="visualization" pkg="mavros_extras" type="visualization" if="$(arg viz)">
        <remap to="mavros/local_position/pose" from="local_position"/>
        <remap to="mavros/setpoint_position/local" from="local_setpoint"/>
        <param name="fixed_frame_id" value="map"/>
        <param name="child_frame_id" value="base_link"/>
        <param name="marker_scale" value="1"/>
        <param name="max_track_size" value="20"/>
        <param name="num_rotors" value="4"/>
    </node>

</launch>